accepted/2020/mono-convergence/platform-specific-httpclient.md

```diff
@@ -0,0 +1,452 @@
+
+**Proposal 2a: ILLinker Magic**
+
+Each SDK would write optional ILLinker custom step which would rewrite construction logic for platform instance with implementation which is controlled by SDK.
+```c#
+// System.Net.Http.dll (for iOS, Android, etc)
+
+public class HttpClientHandler
+{
+  readonly HttpClientHandler platformInstance;
+
+  public HttpClientHandler ()
+  {
+   if (GetType() == type (HttpClientHandler))
+     platformInstance = CreateNewInstance ();
+  }
+
+  // ILinker will recognize this method and change the body to be platform specific
+  private static HttpClientHandler CreateNewInstance ()
+  {
+    throw new PlatformNotSupportedException();    
+  }
+}
+```
+
+**Disadvantages**
+
+* Requires ILLinker to always run which slows down build process.
+* Changing internal state of HttpClientHandler will be tricky as it can break hook used by ILLinker.
+
+**Proposal 2b: Use Reflection**
+
+The construction code would use reflection call to static method for platform where the late binding is required.
+```c#
+// System.Net.Http.dll
+
+public class HttpClientHandler
+{
+  #if iOS
+  [PreserveDependency ("Create", "HttpClientHandlerFactory", "Xamarin.iOS")
+  #endif
+  public HttpClientHandler ()
+  {
+    platformInstance = Type.GetType ("HttpClientHandlerFactory").GetMethod("Create").Invoke () as HttpClientHandler;
+  }
+}
+```
+
+**Disadvantages**
+
+* Costly and slow reflection infrastructure would be pulled in.
+* Need to add PreserveDependency to all possible target assemblies.
+
+**Proposal 2c: Introduce a new System.Net.Http.HandlerFactory.dll assembly**
+
+.NET introduce a new assembly which the only purpose will be to hold the factory contract for the handler construction. All other types will remain in System.Net.Http.dll which will make the build complicated due to circular references.
+```c#
+// System.Net.Http.HandlerFactory.dll
+public static class HttpHandlerFactory
+{
+  // It could be of object type not to take dependency on System.Net.Http
+  public HttpClientHandler Create () => throw new NotImplementedException();
+}
+```
+
+```c#
+// System.Net.Http.dll
+public HttpClientHandler ()
+{
+  platformInstance = HttpHandlerFactory.Create();
+}
+```
+
+**Disadvantages**
+
+* Single method assembly would be added to the shared framework
+* A circular dependency between System.Net.Http.HandlerFactory.dll and System.Net.Http.dll will exist if we use HttpClientHandler as the “common” type.
+
+**Proposal 2d: Add SDK reference assemblies with factory contract to dotnet/runtime**
+
+This is what Mono does today. The reference assemblies which describe the factory method contract will be available during dotnet/runtime build and used for each RID specific build of System.Net.Http.dll.
+```c#
+// System.Net.Http.dll
+public HttpClientHandler ()
+{
+#if iOS
+  platformInstance = Xamarin.iOS.HttpHandleFactory.Create();
+#else
+...
+#endif
+}
+```
```

Comment:
**Disadvantages**

* The assembly `System.Net.Http.dll` is different for each platform

Reply:
> The assembly System.Net.Http.dll is different for each platform

That's already the case.  We build System.Net.Http.dll for Windows, for Linux, and for macOS.  While 99.9% of the code is common between them, there's a tiny bit of code in each that does P/Invokes for the given OS to look up things like proxy settings.